File: src/main/java/com/zuminX/utils/builder/ApiModelGenerator.java
package com.zuminX.utils.builder;

import com.intellij.psi.PsiClass;
import com.zuminX.annotations.swagger.ApiModel;
import com.zuminX.interceptor.AnnotationGeneratorInterceptor;
import com.zuminX.names.ClassName;
import com.zuminX.utils.GeneratorUtils;
import java.util.ArrayList;
import java.util.List;

/**
 * ApiModel注解的生成器类
 */
public class ApiModelGenerator implements AnnotationGenerator<PsiClass, ApiModel> {

  public static final ApiModelGenerator INSTANCE = AnnotationGeneratorInterceptor.create(ApiModelGenerator.class, ApiModel.class);

  /**
   * 根据psi类对象构建ApiModel对象
   *
   * @param psiClass psi类
   * @return ApiModel注解对象
   */
  public final ApiModel build(PsiClass psiClass) {
    return ApiModel.builder()
        .value(getValue(psiClass))
        .description(getDescription(psiClass))
        .parent(getParent(psiClass))
        .subTypes(getSubTypes(psiClass))
        .build();
  }

  /**
   * 从psi类中获取ApiModel类的value属性
   *
   * @param psiClass psi类
   * @return ApiModel类的value属性
   */
  protected String getValue(PsiClass psiClass) {
    return psiClass.getName();
  }

  /**
   * 从psi类中获取ApiModel类的description属性
   *
   * @param psiClass psi类
   * @return ApiModel类的description属性
   */
  protected String getDescription(PsiClass psiClass) {
    return GeneratorUtils.getFirstComment(psiClass);
  }

  /**
   * 从psi类中获取ApiModel类的parent属性
   *
   * @param psiClass psi类
   * @return ApiModel类的parent属性
   */
  protected ClassName getParent(PsiClass psiClass) {
    PsiClass superClass = psiClass.getSuperClass();
    return superClass == null ? ClassName.OBJECT_CLASS_NAME : new ClassName(superClass.getQualifiedName());
  }

  /**
   * 从psi类中获取ApiModel类的subTypes属性
   *
   * @param psiClass psi类
   * @return ApiModel类的subTypes属性
   */
  protected List<ClassName> getSubTypes(PsiClass psiClass) {
    List<ClassName> result = new ArrayList<>();
    PsiClass nowClass = psiClass;
    while (nowClass.getSuperClass() != null) {
      if (!ClassName.OBJECT_CLASS_NAME.getQualifiedName().equals(nowClass.getSuperClass().getQualifiedName())) {
        result.add(getParent(nowClass));
      }
      nowClass = nowClass.getSuperClass();
    }
    // 当且仅当该类无显式父类时，才视为其父类包括Object
    return result.isEmpty() ? List.of(ClassName.OBJECT_CLASS_NAME) : result;
  }
}

File: src/main/java/com/zuminX/generator/builder/ApiModelGenerator.kt
package com.zuminX.generator.builder

import com.intellij.psi.PsiClass
import com.zuminX.annotations.swagger.ApiModel
import com.zuminX.generator.AnnotationGenerator
import com.zuminX.generator.AnnotationGeneratorInterceptor
import com.zuminX.names.ClassName
import com.zuminX.names.OBJECT_CLASS_NAME
import com.zuminX.utils.getFirstComment

val ApiModelGeneratorInstance = AnnotationGeneratorInterceptor.create(ApiModelGenerator::class.java, ApiModel::class.java)

/**
 * ApiModel注解的生成器类
 */
open class ApiModelGenerator : AnnotationGenerator<PsiClass, ApiModel> {

  /**
   * 根据psi类对象构建ApiModel对象
   *
   * @param psiElement psi类
   * @return ApiModel注解对象
   */
  final override fun build(psiElement: PsiClass): ApiModel {
    return ApiModel().apply {
      value = getValue(psiElement)
      description = getDescription(psiElement)
      parent = getParent(psiElement)
      subTypes = getSubTypes(psiElement)
    }
  }

  /**
   * 从psi类中获取ApiModel类的value属性
   *
   * @param psiClass psi类
   * @return ApiModel类的value属性
   */
  protected fun getValue(psiClass: PsiClass): String? {
    return psiClass.name
  }

  /**
   * 从psi类中获取ApiModel类的description属性
   *
   * @param psiClass psi类
   * @return ApiModel类的description属性
   */
  protected fun getDescription(psiClass: PsiClass): String? {
    return psiClass.getFirstComment()
  }

  /**
   * 从psi类中获取ApiModel类的parent属性
   *
   * @param psiClass psi类
   * @return ApiModel类的parent属性
   */
  protected fun getParent(psiClass: PsiClass): ClassName {
    val superClass = psiClass.superClass ?: return OBJECT_CLASS_NAME
    return ClassName(superClass.qualifiedName ?: "")
  }

  /**
   * 从psi类中获取ApiModel类的subTypes属性
   *
   * @param psiClass psi类
   * @return ApiModel类的subTypes属性
   */
  protected fun getSubTypes(psiClass: PsiClass): List<ClassName> {
    val result: MutableList<ClassName> = ArrayList()
    var nowClass: PsiClass = psiClass
    while (nowClass.superClass != null) {
      if (OBJECT_CLASS_NAME.qualifiedName != nowClass.superClass!!.qualifiedName) {
        result.add(getParent(nowClass))
      }
      nowClass = nowClass.superClass!!
    }
    // 当且仅当该类无显式父类时，才视为其父类包括Object
    return if (result.isEmpty()) listOf(OBJECT_CLASS_NAME) else result
  }

}